### PR TITLE
Bump fgbio versions and swap to nftest

### PR DESCRIPTION
**Pull request by SPPearce**
Starting the process of updating fgbio versions.
This is porting the modules from [fastquorum](https://github.com/nf-core/fastquorum), updating them to the most recent fgbio version and swapping to nf-test

### CODE REVIEW

File: modules/nf-core/fgbio/callmolecularconsensusreads/main.nf [hunk 1]
```diff
@@ -19,19 +21,42 @@ process FGBIO_CALLMOLECULARCONSENSUSREADS {
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+    def prefix = task.ext.prefix ?: "${meta.id}_consensus_unmapped"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/callmolecularconsensusreads/main.nf:24]: Why did you add this custom suffix here?

File: modules/nf-core/fgbio/groupreadsbyumi/main.nf [hunk 2]
```diff
@@ -4,35 +4,45 @@ process FGBIO_GROUPREADSBYUMI {
...
     path "versions.yml"                         , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
-
+    def prefix = task.ext.prefix ?: "${meta.id}_umi-grouped"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/groupreadsbyumi/main.nf:24]: Same here? (just curious)

**Review by nvnieuwk — COMMENTED** [review]
I left some comments from a quick look at it, I'll look at it a bit more thoroughly if you fixed these comments :p

File: modules/nf-core/fgbio/callduplexconsensusreads/tests/main.nf.test [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    tag "fgbio/callduplexconsensusreads"
+
+    test("homo_sapiens - bam") {
+
+        when {
+            process {
+                """
+                input[0] = [
+                    [ id:'test', single_end:false ], // meta map
+                    file(params.test_data['homo_sapiens']['illumina']['test_paired_end_duplex_umi_grouped_bam'], checkIfExists: true)
```
[the remark this answers is not in the document]
Comment (SPPearce) [reply modules/nf-core/fgbio/callduplexconsensusreads/tests/main.nf.test:19]: I'm assuming that all modules will get mass-updated at some point

File: modules/nf-core/fgbio/callmolecularconsensusreads/main.nf (hunk 1 above)
Comment (SPPearce) [reply modules/nf-core/fgbio/callmolecularconsensusreads/main.nf:24]: It needs some kind of name renaming to prevent it from colliding with the input file. Some of the subtools already done this, some don't. Although this one is now setting a double prefix in the output file, so that needs fixing.

File: modules/nf-core/fgbio/callmolecularconsensusreads/main.nf [hunk 4]
```diff
@@ -19,19 +21,42 @@ process FGBIO_CALLMOLECULARCONSENSUSREADS {
...
     """
     fgbio \\
+        -Xmx${mem_gb}g \\
         --tmp-dir=. \\
+        --async-io=true \\
+        --compression=1 \\
         CallMolecularConsensusReads \\
-        --input $bam \\
+        --input $grouped_bam \\
+        --output ${prefix}.cons.unmapped.bam \\
```
Comment (SPPearce) [inline modules/nf-core/fgbio/callmolecularconsensusreads/main.nf:39]:
```suggestion
        --output ${prefix}.bam \\
```

File: modules/nf-core/fgbio/callduplexconsensusreads/tests/main.nf.test (hunk 3 above)
[the remark this answers is not in the document]
Comment (nvnieuwk) [reply modules/nf-core/fgbio/callduplexconsensusreads/tests/main.nf.test:19]: I'm not sure it will be that easy... It's also in the guidelines now (https://nf-co.re/docs/guidelines/components/modules#input-data) :)

File: modules/nf-core/fgbio/callmolecularconsensusreads/main.nf (hunk 1 above)
Comment (nvnieuwk) [reply modules/nf-core/fgbio/callmolecularconsensusreads/main.nf:24]: You can add a check for that (a lot of modules do this already) => https://nf-co.re/docs/guidelines/components/modules#command-file-output-naming

Comment (SPPearce) [reply modules/nf-core/fgbio/callmolecularconsensusreads/main.nf:24]: Yes, it needs the checks whatever.
I think it is better to have a default prefix that doesn't require you to add one.

File: modules/nf-core/fgbio/callduplexconsensusreads/tests/main.nf.test (hunk 3 above)
[the remark this answers is not in the document]
Comment (SPPearce) [reply modules/nf-core/fgbio/callduplexconsensusreads/tests/main.nf.test:19]: Mmm, I've merged like 50+ PRs in the last few weeks with the old structure.
But I guess if the guidelines explicitly say that now I guess I'll change it.

File: modules/nf-core/fgbio/callmolecularconsensusreads/main.nf (hunk 1 above)
Comment (nvnieuwk) [reply modules/nf-core/fgbio/callmolecularconsensusreads/main.nf:24]: I'm not one to argue with that, but I don't think diverting from the guidelines is a good idea, they're here for a reason. You can maybe bring this up on Slack?

File: modules/nf-core/fgbio/callmolecularconsensusreads/main.nf [hunk 5]
```diff
@@ -19,19 +21,44 @@ process FGBIO_CALLMOLECULARCONSENSUSREADS {
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+    def prefix = task.ext.prefix ?: "${meta.id}_consensus_unmapped"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/callmolecularconsensusreads/main.nf:24]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/nf-core/fgbio/callmolecularconsensusreads/main.nf [hunk 6]
```diff
@@ -19,19 +21,44 @@ process FGBIO_CALLMOLECULARCONSENSUSREADS {
...
+        $args;
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         fgbio: \$( echo \$(fgbio --version 2>&1 | tr -d '[:cntrl:]' ) | sed -e 's/^.*Version: //;s/\\[.*\$//')
     END_VERSIONS
     """
+
+    stub:
+    prefix = task.ext.prefix ?: "${meta.id}_consensus_unmapped"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/callmolecularconsensusreads/main.nf:53]:
```suggestion
    prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/nf-core/fgbio/groupreadsbyumi/main.nf [hunk 7]
```diff
@@ -4,35 +4,60 @@ process FGBIO_GROUPREADSBYUMI {
...
+    tuple val(meta), path("*histogram.txt"), emit: histogram
+    path "versions.yml"                    , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+    def prefix = task.ext.prefix ?: "${meta.id}_umi-grouped"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/groupreadsbyumi/main.nf:24]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/nf-core/fgbio/groupreadsbyumi/main.nf [hunk 8]
```diff
@@ -4,35 +4,60 @@ process FGBIO_GROUPREADSBYUMI {
...
+        -f ${prefix}_histogram.txt
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        fgbio: \$( echo \$(fgbio --version 2>&1 | tr -d '[:cntrl:]' ) | sed -e 's/^.*Version: //;s/\\[.*\$//')
+    END_VERSIONS
+    """
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}_umi-grouped"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/groupreadsbyumi/main.nf:56]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/nf-core/fgbio/sortbam/main.nf [hunk 9]
```diff
@@ -19,9 +19,23 @@ process FGBIO_SORTBAM {
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+    def prefix = task.ext.prefix ?: "${meta.id}_sorted"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/sortbam/main.nf:22]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/nf-core/fgbio/sortbam/main.nf [hunk 10]
```diff
@@ -32,5 +46,17 @@ process FGBIO_SORTBAM {
         fgbio: \$( echo \$(fgbio --version 2>&1 | tr -d '[:cntrl:]' ) | sed -e 's/^.*Version: //;s/\\[.*\$//')
     END_VERSIONS
     """
+
+    stub:
+    prefix = task.ext.prefix ?: "${meta.id}_sorted"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/sortbam/main.nf:51]:
```suggestion
    prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/nf-core/fgbio/zipperbams/main.nf [hunk 11]
```diff
@@ -25,20 +25,22 @@ process FGBIO_ZIPPERBAMS {
     def args2 = task.ext.args2 ?: ''
     def compression = task.ext.compression ?: '0'
     prefix = task.ext.prefix ?: "${meta.id}_zipped"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/zipperbams/main.nf:27]:
```suggestion
    prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/nf-core/fgbio/zipperbams/main.nf [hunk 12]
```diff
@@ -49,6 +51,20 @@ process FGBIO_ZIPPERBAMS {
...
 
 
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        fgbio: \$( echo \$(fgbio --version 2>&1 | tr -d '[:cntrl:]' ) | sed -e 's/^.*Version: //;s/\\[.*\$//')
+    END_VERSIONS
+    """
+
+    stub:
+    prefix = task.ext.prefix ?: "${meta.id}_zipped"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/zipperbams/main.nf:61]:
```suggestion
    prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/nf-core/fgbio/callduplexconsensusreads/main.nf [hunk 13]
```diff
@@ -4,26 +4,24 @@ process FGBIO_CALLDUPLEXCONSENSUSREADS {
...
-    path "versions.yml"                     , emit: versions
+    path "versions.yml"                   , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
 
     script:
     def args = task.ext.args ?: ''
-    prefix = task.ext.prefix ?: "${meta.id}_consensus"
+    prefix = task.ext.prefix ?: "${meta.id}_consensus_unmapped"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/callduplexconsensusreads/main.nf:24]:
```suggestion
    prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/nf-core/fgbio/callduplexconsensusreads/main.nf [hunk 14]
```diff
@@ -53,4 +54,17 @@ process FGBIO_CALLDUPLEXCONSENSUSREADS {
         fgbio: \$( echo \$(fgbio --version 2>&1 | tr -d '[:cntrl:]' ) | sed -e 's/^.*Version: //;s/\\[.*\$//')
     END_VERSIONS
     """
+
+    stub:
+    prefix = task.ext.prefix ?: "${meta.id}_consensus"
```
Comment (nvnieuwk) [inline modules/nf-core/fgbio/callduplexconsensusreads/main.nf:59]:
```suggestion
    prefix = task.ext.prefix ?: "${meta.id}"
```

**Review by nvnieuwk — APPROVED** [review]
Hi some suggestions left (I was unable to provide suggestions for `fgbio/fastqtobam`, but can you fix that one too?). Feel free to merge once you've addressed those :)

**Comment by SPPearce** [comment]
As is currently implemented here, I am keen to include both the check that the input and output files have different names, but also set a default value on the prefix.

File: modules/nf-core/fgbio/callduplexconsensusreads/main.nf (hunk 13 above)
Comment (nh13) [reply modules/nf-core/fgbio/callduplexconsensusreads/main.nf:24]: I am curious why remove the suffix?  This can cause name collisions when naively publishing, as well as I think the suffix makes the file a lot more descriptive.

File: modules/nf-core/fgbio/callmolecularconsensusreads/meta.yml [hunk 15]
```diff
@@ -17,11 +16,17 @@ input:
       description: |
         Groovy Map containing sample information
         e.g. [ id:'test', single_end:false, collapse:false ]
-  - bam:
+  - grouped_bam:
```
Comment (nh13) [inline modules/nf-core/fgbio/callmolecularconsensusreads/meta.yml:19]: ditto, does the `main.nf` need to be updated?

File: modules/nf-core/fgbio/groupreadsbyumi/meta.yml [hunk 16]
```diff
@@ -29,7 +29,7 @@ input:
         BAM file. Note: the MQ tag is required on reads with mapped mates (!)
       pattern: "*.bam"
   - strategy:
-      type: value
+      type: string
```
Comment (nh13) [inline modules/nf-core/fgbio/groupreadsbyumi/meta.yml:32]:
```suggestion
      type: string
      enum: ["Identity", "Edit", "Adjacency", "Paired"]
```

File: modules/nf-core/fgbio/zipperbams/meta.yml [hunk 17]
```diff
@@ -16,28 +19,51 @@ input:
...
+        e.g. [ id:'test', single_end:false ]
+  - unmapped_bam:
       type: file
       description: unmapped BAM file
       pattern: "*.bam"
+  - meta3:
+      type: map
+      description: |
+        Groovy Map containing reference information
+        e.g. [ id:'test', single_end:false ]
```
Comment (nh13) [inline modules/nf-core/fgbio/zipperbams/meta.yml:39]:
```suggestion
        e.g. [ id:'GRCh38' ]
```

File: modules/nf-core/fgbio/zipperbams/meta.yml [hunk 18]
```diff
@@ -16,28 +19,51 @@ input:
...
+  - fasta:
+      type: file
+      description: fasta file containing genomic sequence information
+      pattern: "*.{fasta,fa}"
+  - meta4:
+      type: map
+      description: |
+        Groovy Map containing reference information
+        e.g. [ id:'test', single_end:false ]
```
Comment (nh13) [inline modules/nf-core/fgbio/zipperbams/meta.yml:48]:
```suggestion
        e.g. [ id:'GRCh38' ]
```

File: modules/nf-core/fgbio/callduplexconsensusreads/main.nf (hunk 13 above)
Comment (nvnieuwk) [reply modules/nf-core/fgbio/callduplexconsensusreads/main.nf:24]: The guidelines state that the default `prefix` should be the `meta.id`. It's up to the pipeline developer to make the prefix more descriptive since this can differ from pipeline to pipeline.

Comment (nh13) [reply modules/nf-core/fgbio/callduplexconsensusreads/main.nf:24]: Gotcha. . I’d argue this is a bit of a foot gun, but if the _“Guidelines”_ would stop this PR from being merged, I’d understand.

Comment (SPPearce) [reply modules/nf-core/fgbio/callduplexconsensusreads/main.nf:24]: I would definitely agree that it is a foot gun, and I'm working on altering the guidelines ;)

Comment (lescai) [reply modules/nf-core/fgbio/callduplexconsensusreads/main.nf:24]: this usually goes in the config, and it should be set this way on the nf-test config. 
@nh13 highlights a quite painful point, which we probably haven't succeeded documenting well for the casual user, i.e. that test configs represent an example for implementation :)

File: modules/nf-core/fgbio/callduplexconsensusreads/meta.yml [hunk 19]
```diff
@@ -10,18 +10,23 @@ tools:
...
+      licence: ["MIT"]
 input:
-  # Only when we have meta
   - meta:
       type: map
       description: |
         Groovy Map containing sample information
         e.g. [ id:'test', single_end:false ]
-  - bam:
+  - grouped_bam:
```
[the remark this answers is not in the document]
Comment (SPPearce) [reply modules/nf-core/fgbio/callduplexconsensusreads/meta.yml:20]: The tuple is flattened in this list.
It does need to be consistent between the main.nf and meta.yml though. I changed them all to be `bam` for consistency (although them being grouped_bam does make it clear what the input needs to be from the code rather than the meta.yml)

Comment (SPPearce) [inline modules/nf-core/fgbio/callduplexconsensusreads/meta.yml:20]:
```suggestion
  - bam:
```
